src/common/faucet.ts

```diff
@@ -0,0 +1,234 @@
+}
+
+export class ARIOTokenFaucet implements TokenFaucet {
+  private faucetUrl: string;
+  private processId: string;
+
+  constructor({
+    faucetUrl,
+    processId,
+  }: {
+    faucetUrl: string;
+    processId: string;
+  }) {
+    this.faucetUrl = faucetUrl;
+    this.processId = processId;
+  }
+
+  /**
+   * Returns the captcha URL for a process. The captcha is used to verify a human is solving the captcha. Once you have a captcha response, you can use it to request an authorization token via the requestAuthToken method.
+   * @returns The captcha URL for a process
+   */
+  async captchaUrl(): Promise<{
+    processId: string;
+    captchaUrl: string;
+  }> {
+    const res = await fetch(
+      `${this.faucetUrl}/api/captcha/url?process-id=${this.processId}`,
+      {
+        method: 'GET',
+      },
+    );
+
+    if (!res.ok) {
+      const body = await res.json().catch(() => ({ error: res.statusText }));
+      throw new Error((body as { error: string }).error);
+    }
+    const data = (await res.json()) as {
+      processId: string;
+      captchaUrl: string;
+    };
+    return data;
+  }
+
+  /**
+   * Claim tokens for a process using a captcha response. This method is used to synchronously claim tokens for a process using a captcha response.
+   * @param captchaResponse - The captcha response
+   * @param recipient - The recipient address
+   * @param quantity - The quantity of tokens to claim
+   * @returns The claim id and success status
+   */
+  async claimWithCaptchaResponse({
+    captchaResponse,
+    recipient,
+    quantity,
+  }: {
+    captchaResponse: string;
+    recipient: string;
+    quantity: number;
+  }): Promise<{ id: string; success: boolean }> {
+    const res = await fetch(`${this.faucetUrl}/api/claim/sync`, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({
+        processId: this.processId,
+        recipient,
+        quantity,
+        captchaResponse,
+      }),
+    });
+
+    if (!res.ok) {
+      const body = await res.json().catch(() => ({ error: res.statusText }));
+      throw new Error((body as { error: string }).error);
+    }
+
+    const data = (await res.json()) as { id: string; success: boolean };
+    return data;
+  }
+
+  /**
+   * Requests an authorization token for a process. The captcha response is used to verify a human is solving the captcha. Once you have an authorization token, you can use it to claim tokens from the faucet via the claimWithAuthToken method.
+   * @param captchaResponse - The captcha response
+   * @returns The status of the request, the authorization token, and the expiration time
+   */
+  async requestAuthToken({
+    captchaResponse,
+  }: {
+    captchaResponse: string;
+  }): Promise<{
+    status: 'success' | 'error';
+    token: string;
+    expiresAt: number;
+  }> {
+    const res = await fetch(`${this.faucetUrl}/api/captcha/verify`, {
```

Comment:
nit: would be nice to have a general retry strategy around node fetch for network issues